File: agent/src/provider/openwrt.py
import re
import yaml
from src.provider.debian import Debian
from src.utils.get_properties import get_properties
from src.grpc.agent_pb2 import AgentApp, AgentWebItem, AgentOS
from src.provider.provider import Provider


class OpenWRT(Provider):
    def check(self):
        if not self.exec.isFile("/etc/glversion"):
            return False

        osReleasePath = "/etc/os-release"
        if not self.exec.isFile(osReleasePath):
            return False

        config_file = self.exec.open(osReleasePath)
        configs = get_properties(config_file)
        osName = configs["NAME"]
        return bool(osName and "openwrt" in osName.lower())

    def get_os(self):
        return AgentOS.OPENWRT_GLINET

    def get_lan(self):
        return Debian(self.exec).get_lan()

    def get_wan(self):
        return Debian(self.exec).get_wan()

    def get_ddns(self):
        ddnsFilePath = "/etc/config/gl_ddns"
        if not self.exec.isFile(ddnsFilePath):
            return None

        ddnsFile = self.exec.open(ddnsFilePath)

        enabled = False
        domain = None

        for lineRaw in ddnsFile.splitlines():
            line = lineRaw.strip()
            if line.startswith("config service"):
                enabled = False
                domain = None

            if line.startswith("option domain"):
                domain = line.split(" ")[2].replace("'", "")

            if line.startswith("option enabled"):
                enabled = bool(int(line.split(" ")[2].replace("'", "")))

            if enabled and domain:
                return domain

        if enabled and domain:
            return domain
        return None

    def get_dhcp(self):
        return Debian(self.exec).get_dhcp()

    def get_web(self) -> list[AgentWebItem]:
        # extract from nginx too complex (lua scripts)
        # TODO user defined
        return []

    def get_ssh(self):
        return Debian(self.exec).get_ssh()

    def get_apps(self) -> list[AgentApp]:
        return list(
            filter(
                None,
                [
                    self.get_wireguard(),
                    self.get_nginx(),
                    self.get_adguard_home(),
                ],
            )
        )

    def get_wireguard(self) -> AgentApp | None:
        configFilePath = "/etc/config/wireguard_server"
        if not self.exec.isFile(configFilePath):
            return None

        configFile = self.exec.open(configFilePath)

        vpnAddress = ""
        vpnClients: list[str] = []

        for lineRaw in configFile.splitlines():
            line = lineRaw.strip()
            lineParts = line.split(" ")

            if lineParts[0] == "option":
                optionValue = lineParts[2].replace("'", "")
                if lineParts[1] == "address_v4":
                    vpnAddress = optionValue.split("/")[0]

                if lineParts[1] == "client_ip":
                    vpnClients.append(optionValue.split("/")[0])

        return AgentApp(
            slug=AgentApp.AgentAppSlug.WIREGUARD,
            vpnMode=AgentApp.AgentVpnMode.SERVER,
            vpnAddress=vpnAddress,
            vpnClients=vpnClients,
        )

    def get_nginx(self) -> AgentApp | None:
        reverseProxy: list[AgentApp.AgentReverseProxy] = []

        configFolderPath = "/etc/nginx"
        if not self.exec.isDir(configFolderPath):
            return None

        configFilesPaths = self.exec.exec(
            f"find {configFolderPath} -type f -name '*.conf'"
        ).splitlines()

        configFiles: list[str] = []

        for path in configFilesPaths:
            content = self.exec.open(path)
            configFiles.append(content)

        pattern = re.compile(
            r"[^#]*?listen\s+(\d+)[^#]*?server_name\s+([^\s]+)\s*;[^#]*?proxy_pass\s+([^\s]+);",
            re.DOTALL,
        )

        for fileContent in configFiles:
            matches: list[tuple[str, str, str]] = pattern.findall(fileContent)
            for match in matches:
                fromPort, domain, fullAddress = match
                addressParts = fullAddress.split("://")
                ssl = addressParts[0] == "https"
                address, port = addressParts[1].split(":")
                reverseProxy.append(
                    AgentApp.AgentReverseProxy(
                        fromDomain=AgentApp.AgentReverseProxy.From(
                            domain=domain,
                            ssl=int(fromPort) == 443,
                        ),
                        toAddress=AgentApp.AgentReverseProxy.To(
                            address=address, ssl=ssl, port=int(port) if port else None
                        ),
                    )
                )

        # print(reverseProxy)

        return AgentApp(slug=AgentApp.AgentAppSlug.NGINX, reverseProxy=reverseProxy)

    def get_adguard_home(self) -> AgentApp | None:
        configFilePath = "/etc/AdGuardHome/config.yaml"
        if not self.exec.isFile(configFilePath):
            return None

        config = yaml.safe_load(self.exec.open(configFilePath))

        web: list[AgentWebItem] = []

        if "http" in config and "address" in config["http"]:
            address: str = config["http"]["address"]
            port = address.split(":")[1]
            web.append(AgentWebItem(port=int(port)))

        return AgentApp(slug=AgentApp.AgentAppSlug.ADGUARD_HOME, web=web)

File: agent/src/provider/openwrt.go
package provider

import (
	ex "agent/src/executor"
	gen "agent/src/grpc/generated"
	utils "agent/src/utils"
	"regexp"
	"slices"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
)

type OpenWRTProvider struct {
	*BaseProvider
}

func NewOpenWRTProvider(executor ex.Executor) *OpenWRTProvider {
	return &OpenWRTProvider{
		BaseProvider: NewProvider(executor),
	}
}

func (e *OpenWRTProvider) Check() bool {
	if !e.executor.IsFile("/etc/glversion") {
		return false
	}

	osReleasePath := "/etc/os-release"
	if !e.executor.IsFile(osReleasePath) {
		return false
	}

	configFile := e.executor.Open(osReleasePath)
	configs := utils.GetProperties(configFile)
	osName, hasOsName := configs["NAME"]
	return hasOsName && strings.Contains(strings.ToLower(osName), "openwrt")
}

func (l *OpenWRTProvider) GetInstance() gen.AgentInstance {
	return l.BaseProvider.GetInstance(l)
}

func (e *OpenWRTProvider) GetOS() gen.AgentOS {
	return gen.AgentOS_OPENWRT_GLINET
}

func (e *OpenWRTProvider) GetLan() string {
	return NewDebianProvider(e.executor).GetLan()
}

func (e *OpenWRTProvider) GetWan() *string {
	return NewDebianProvider(e.executor).GetWan()
}

func (e *OpenWRTProvider) GetDDNS() *string {
	ddnsFilePath := "/etc/config/gl_ddns"
	if !e.executor.IsFile(ddnsFilePath) {
		return nil
	}

	ddnsFile := e.executor.Open(ddnsFilePath)

	enabled := false
	domain := ""

	for _, lineRaw := range strings.Split(ddnsFile, "\n") {
		line := strings.TrimSpace(lineRaw)
		if strings.HasPrefix(line, "config service") {
			enabled = true
			domain = ""
		}

		if strings.HasPrefix(line, "option domain") {
			domain = strings.ReplaceAll(strings.Split(line, " ")[2], "'", "")
		}

		if strings.HasPrefix(line, "option enabled") {
			enabledStr := strings.ReplaceAll(strings.Split(line, " ")[2], "'", "")
			enabled = enabledStr == "1"
		}

		if enabled && domain != "" {
			return &domain
		}
	}

	if enabled && domain != "" {
		return &domain
	}
	return nil
}

func (e *OpenWRTProvider) GetDHCP() []*gen.AgentInstance_AgentDHCPItem {
	return NewDebianProvider(e.executor).GetDHCP()
}

func (a *OpenWRTProvider) GetWeb() []*gen.AgentWebItem {
	var webList []*gen.AgentWebItem

	return webList
}

func (a *OpenWRTProvider) GetSSH() *gen.AgentSSH {
	return NewDebianProvider(a.executor).GetSSH()
}

func (a *OpenWRTProvider) GetApps() []*gen.AgentApp {
	return slices.DeleteFunc(
		[]*gen.AgentApp{
			a.GetWireguard(),
			a.GetNginx(),
			a.GetAdguardHome(),
		},
		func(app *gen.AgentApp) bool {
			return app == nil
		},
	)
}

func (e *OpenWRTProvider) GetInstanceList() []*gen.AgentInstance {
	return NewDebianProvider(e.executor).GetInstanceList()
}

func (a *OpenWRTProvider) GetWireguard() *gen.AgentApp {
	configFilePath := "/etc/config/wireguard_server"
	if !a.executor.IsFile(configFilePath) {
		return nil
	}

	configFile := a.executor.Open(configFilePath)

	vpnAddress := ""
	var vpnClients []string

	for _, lineRaw := range strings.Split(configFile, "\n") {
		line := strings.TrimSpace(lineRaw)
		parts := strings.Fields(line)

		if len(parts) < 3 || parts[0] != "option" {
			continue
		}

		optionValue := strings.ReplaceAll(parts[2], "'", "")
		switch parts[1] {
		case "address_v4":
			vpnAddress = strings.Split(optionValue, "/")[0]
		case "client_ip":
			vpnClients = append(vpnClients, strings.Split(optionValue, "/")[0])
		}
	}

	if vpnAddress == "" {
		return nil
	}

	var agentVpnModeServer gen.AgentApp_AgentVpnMode = gen.AgentApp_SERVER

	return &gen.AgentApp{
		Slug:       gen.AgentApp_WIREGUARD,
		VpnMode:    &agentVpnModeServer,
		VpnAddress: &vpnAddress,
		VpnClients: vpnClients,
	}
}

func (a *OpenWRTProvider) GetNginx() *gen.AgentApp {
	var reverseProxy []*gen.AgentApp_AgentReverseProxy

	configFolderPath := "/etc/nginx"
	if !a.executor.IsDir(configFolderPath) {
		return nil
	}

	output, _ := a.executor.Exec("find " + configFolderPath + " -type f -name '*.conf'")
	configFilesPaths := strings.Split(output, "\n")

	configFiles := []string{}

	for _, path := range configFilesPaths {
		content := a.executor.Open(path)
		configFiles = append(configFiles, content)
	}

	pattern := regexp.MustCompile(`(?s)[^#]*?listen\s+(\d+)[^#]*?server_name\s+([^\s]+)\s*;[^#]*?proxy_pass\s+([^\s]+);`)

	for _, fileContent := range configFiles {
		matches := pattern.FindAllStringSubmatch(fileContent, -1)

		for _, match := range matches {
			if len(match) < 4 {
				continue
			}

			fromPort := match[1]
			domain := match[2]
			fullAddress := match[3]

			addressParts := strings.Split(fullAddress, "://")
			ssl := addressParts[0] == "https"
			toParts := strings.Split(addressParts[1], ":")

			var port32 int32
			if len(toParts) > 1 {
				port, _ := strconv.Atoi(toParts[1])
				port32 = int32(port)
			}

			reverseProxy = append(reverseProxy, &gen.AgentApp_AgentReverseProxy{
				FromDomain: &gen.AgentApp_AgentReverseProxy_From{
					Domain: domain,
					Ssl:    fromPort == "443",
				},
				ToAddress: &gen.AgentApp_AgentReverseProxy_To{
					Address: toParts[0],
					Ssl:     ssl,
					Port:    &port32,
				},
			})
		}
	}

	if len(reverseProxy) == 0 {
		return nil
	}

	return &gen.AgentApp{
		Slug:         gen.AgentApp_NGINX,
		ReverseProxy: reverseProxy,
	}
}

func (a *OpenWRTProvider) GetAdguardHome() *gen.AgentApp {
	configFilePath := "/etc/AdGuardHome/config.yaml"
	if !a.executor.IsFile(configFilePath) {
		return nil
	}

	content := a.executor.Open(configFilePath)
	var config map[string]interface{}
	yaml.Unmarshal([]byte(content), &config)

	var web []*gen.AgentWebItem
	if httpConfig, ok := config["http"].(map[string]interface{}); ok {
		if address, ok := httpConfig["address"].(string); ok {
			parts := strings.Split(address, ":")
			if len(parts) > 1 {
				port, _ := strconv.Atoi(parts[1])
				port32 := int32(port)
				web = append(web, &gen.AgentWebItem{Port: &port32})
			}
		}
	}

	return &gen.AgentApp{
		Slug: gen.AgentApp_ADGUARD_HOME,
		Web:  web,
	}
}
